<!DOCTYPE html><html class=writer-html5 lang=en> <head><meta charset=utf-8><meta http-equiv=X-UA-Compatible content="IE=edge"><meta name=viewport content="width=device-width, initial-scale=1.0"><meta name=author content="Wilfried Guiblet, Ph.D. and Vishal Koparde, Ph.D."><link rel="shortcut icon" href=../../img/favicon.ico><title>Run - NCI-RBL Dockers</title><link rel=stylesheet href=../../css/theme.css><link rel=stylesheet href=../../css/theme_extra.css><link rel=stylesheet href=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css><script>
        // Current page data
        var mkdocs_page_name = "Run";
        var mkdocs_page_input_path = "LEAFCUTTER/run.md";
        var mkdocs_page_url = null;
      </script><!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]--><script src=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js></script><script>hljs.highlightAll();</script></head> <body class=wy-body-for-nav role=document> <div class=wy-grid-for-nav> <nav data-toggle=wy-nav-shift class="wy-nav-side stickynav"> <div class=wy-side-scroll> <div class=wy-side-nav-search> <a href=../.. class="icon icon-home"> NCI-RBL Dockers </a><div role=search> <form id=rtd-search-form class=wy-form action=../../search.html method=get> <input type=text name=q placeholder="Search docs" aria-label="Search docs" title="Type search term here"> </form> </div> </div> <div class="wy-menu wy-menu-vertical" data-spy=affix role=navigation aria-label="Navigation menu"> <ul> <li class=toctree-l1><a class="reference internal" href=../..>Intro</a> </li> </ul> <p class=caption><span class=caption-text>MOP2</span></p> <ul> <li class=toctree-l1><a href=../../MOP2/descript/ class="reference internal">Description</a> </li> <li class=toctree-l1><a href=../../MOP2/init/ class="reference internal">Initialize</a> </li> <li class=toctree-l1><a href=../../MOP2/run/ class="reference internal">Run</a> </li> </ul> <p class=caption><span class=caption-text>DTEG</span></p> <ul> <li class=toctree-l1><a href=../../DTEG/descript/ class="reference internal">Description</a> </li> <li class=toctree-l1><a href=../../DTEG/init/ class="reference internal">Initialize</a> </li> <li class=toctree-l1><a href=../../DTEG/run/ class="reference internal">Run</a> </li> </ul> <p class=caption><span class=caption-text>ISOMIR</span></p> <ul> <li class=toctree-l1><a href=../../ISOMIR/descript/ class="reference internal">Description</a> </li> <li class=toctree-l1><a href=../../ISOMIR/init/ class="reference internal">Initialize</a> </li> <li class=toctree-l1><a href=../../ISOMIR/run/ class="reference internal">Run</a> </li> </ul> <p class=caption><span class=caption-text>RiboFootPrint</span></p> <ul> <li class=toctree-l1><a href=../../RiboFootPrint/descript/ class="reference internal">Description</a> </li> <li class=toctree-l1><a href=../../RiboFootPrint/init/ class="reference internal">Initialize</a> </li> <li class=toctree-l1><a href=../../RiboFootPrint/run/ class="reference internal">Run</a> </li> </ul> <p class=caption><span class=caption-text>iCLIP</span></p> <ul> <li class=toctree-l1><a href=../../iCLIP/descript/ class="reference internal">Description</a> </li> <li class=toctree-l1><a href=../../iCLIP/init/ class="reference internal">Initialize</a> </li> <li class=toctree-l1><a href=../../iCLIP/run/ class="reference internal">Run</a> </li> </ul> <p class=caption><span class=caption-text>eCLIP</span></p> <ul> <li class=toctree-l1><a href=../../eCLIP/eCLIP_workflow/ class="reference internal">Workflow</a> </li> </ul> <p class=caption><span class=caption-text>LEAFCUTTER</span></p> <ul class=current> <li class=toctree-l1><a href=../descript/ class="reference internal">Description</a> </li> <li class=toctree-l1><a href=../init/ class="reference internal">Initialize</a> </li> <li class="toctree-l1 current"><a class="reference internal current" href=#>Run</a> <ul class=current> </ul> </li> </ul> </div> </div> </nav> <section data-toggle=wy-nav-shift class=wy-nav-content-wrap> <nav class=wy-nav-top role=navigation aria-label="Mobile navigation menu"> <i data-toggle=wy-nav-top class="fa fa-bars"></i> <a href=../..>NCI-RBL Dockers</a> </nav> <div class=wy-nav-content> <div class=rst-content><div role=navigation aria-label="breadcrumbs navigation"> <ul class=wy-breadcrumbs> <li><a href=../.. class="icon icon-home" aria-label=Docs></a></li> <li class=breadcrumb-item>LEAFCUTTER</li> <li class="breadcrumb-item active">Run</li> <li class=wy-breadcrumbs-aside> <a href=https://github.com/NCI-RBL/Dockers/edit/main/docs/LEAFCUTTER/run.md>Edit on NCI-RBL/Dockers</a> </li> </ul> <hr> </div> <div role=main class=document itemscope=itemscope itemtype=http://schema.org/Article> <div class=section itemprop=articleBody> <h3 id=run-star>Run STAR:<a class=headerlink href=#run-star title="Permanent link">&para;</a></h3> <p>Align your samples in twopassMode for more accurate junction calling:</p> <div class=highlight><pre><span></span><code>STAR<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--runThreadN<span class=w> </span><span class=m>64</span><span class=w> </span><span class=se>\</span>
<span class=w>  </span>--runMode<span class=w> </span>alignReads<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--genomeDir<span class=w> </span><span class=s1>&#39;STAR_hg38&#39;</span><span class=w> </span><span class=se>\</span>
<span class=w>  </span>--sjdbGTFfile<span class=w> </span><span class=s1>&#39;gencode.v44.annotation.gtf&#39;</span><span class=w> </span><span class=se>\</span>
<span class=w>  </span>--outFileNamePrefix<span class=w> </span><span class=si>${</span><span class=nv>sample</span><span class=si>}</span>_<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--outSAMunmapped<span class=w> </span>Within<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--outFilterType<span class=w> </span>BySJout<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--outFilterMultimapNmax<span class=w> </span><span class=m>20</span><span class=w> </span><span class=se>\</span>
<span class=w>  </span>--outFilterMismatchNmax<span class=w> </span><span class=m>999</span><span class=w> </span><span class=se>\</span>
<span class=w>  </span>--outFilterMismatchNoverLmax<span class=w> </span><span class=m>0</span>.04<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--alignIntronMin<span class=w> </span><span class=m>20</span><span class=w> </span><span class=se>\</span>
<span class=w>  </span>--alignIntronMax<span class=w> </span><span class=m>50000</span><span class=w> </span><span class=se>\ </span>
<span class=w>  </span>--alignMatesGapMax<span class=w> </span><span class=m>1000000</span><span class=w> </span><span class=se>\</span>
<span class=w>  </span>--alignSJoverhangMin<span class=w> </span><span class=m>5</span><span class=w> </span><span class=se>\</span>
<span class=w>  </span>--alignSJDBoverhangMin<span class=w> </span><span class=m>3</span><span class=w> </span><span class=se>\</span>
<span class=w>  </span>--sjdbScore<span class=w> </span><span class=m>1</span><span class=w> </span><span class=se>\</span>
<span class=w>  </span>--readFilesIn<span class=w> </span><span class=si>${</span><span class=nv>sample</span><span class=si>}</span>_R1.fq<span class=w> </span><span class=se>\</span>
<span class=w>                </span><span class=si>${</span><span class=nv>sample</span><span class=si>}</span>_R2.fq<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--outFilterMatchNminOverLread<span class=w> </span><span class=m>0</span>.66<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--outSAMtype<span class=w> </span>BAM<span class=w> </span>Unsorted<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--quantMode<span class=w> </span>TranscriptomeSAM<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--peOverlapNbasesMin<span class=w> </span><span class=m>10</span><span class=w> </span><span class=se>\</span>
<span class=w>  </span>--alignEndsProtrude<span class=w> </span><span class=m>10</span><span class=w> </span>ConcordantPair<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--twopassMode<span class=w> </span>Basic<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--outSAMstrandField<span class=w> </span>intronMotif<span class=w> </span>

samtools<span class=w> </span>sort<span class=w> </span>-@<span class=w> </span><span class=m>64</span><span class=w> </span>-o<span class=w> </span><span class=si>${</span><span class=nv>sample</span><span class=si>}</span>_Aligned.sorted.bam<span class=w> </span><span class=si>${</span><span class=nv>sample</span><span class=si>}</span>_Aligned.out.bam
samtools<span class=w> </span>index<span class=w> </span>-@<span class=w> </span><span class=m>64</span><span class=w> </span>-M<span class=w> </span><span class=si>${</span><span class=nv>sample</span><span class=si>}</span>_Aligned.sorted.bam<span class=w> </span>
</code></pre></div> <h3 id=run-regtools>Run regtools<a class=headerlink href=#run-regtools title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><span class=k>for</span><span class=w> </span>bamfile<span class=w> </span><span class=k>in</span><span class=w> </span><span class=sb>`</span>ls<span class=w> </span>*_Aligned.sorted.bam<span class=sb>`</span><span class=p>;</span><span class=w> </span><span class=k>do</span>
<span class=w>    </span><span class=nb>echo</span><span class=w> </span>Converting<span class=w> </span><span class=nv>$bamfile</span><span class=w> </span>to<span class=w> </span><span class=nv>$bamfile</span>.junc
<span class=w>    </span>regtools<span class=w> </span>junctions<span class=w> </span>extract<span class=w> </span>-s<span class=w> </span>XS<span class=w> </span>-a<span class=w> </span><span class=m>8</span><span class=w> </span>-m<span class=w> </span><span class=m>50</span><span class=w> </span>-M<span class=w> </span><span class=m>500000</span><span class=w> </span><span class=nv>$bamfile</span><span class=w> </span>-o<span class=w> </span><span class=nv>$bamfile</span>.junc
<span class=k>done</span>
</code></pre></div> <p>Repeat alignment for all samples. Leafcutter documentation recommends 4 samples per group mimimum.</p> <h3 id=download-and-run-leafcutter-container>Download and Run leafcutter container<a class=headerlink href=#download-and-run-leafcutter-container title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><span class=nb>export</span><span class=w> </span><span class=nv>SINGULARITY_CACHEDIR</span><span class=o>=</span><span class=nv>$PWD</span>/.singularity
singularity<span class=w> </span>pull<span class=w> </span>--dir<span class=w> </span>./<span class=w> </span>leafcutter.sif<span class=w> </span>docker://wilfriedguiblet/leafcutter:v0.1
</code></pre></div> <p>Create a juncfiles.txt. Example: <div class=highlight><pre><span></span><code>Sample1_Aligned.sorted.bam.junc
Sample2_Aligned.sorted.bam.junc
Sample3_Aligned.sorted.bam.junc
Sample4_Aligned.sorted.bam.junc
Sample5_Aligned.sorted.bam.junc
Sample6_Aligned.sorted.bam.junc
Sample7_Aligned.sorted.bam.junc
Sample8_Aligned.sorted.bam.junc
</code></pre></div></p> <p>Download custom leafcutter:</p> <p>git clone <a href=https://github.com/wilfriedguiblet/leafcutter.git>https://github.com/wilfriedguiblet/leafcutter.git</a></p> <p>Enter the container <div class=highlight><pre><span></span><code>singularity<span class=w> </span>shell<span class=w> </span>-B<span class=w> </span><span class=k>$(</span><span class=nb>pwd</span><span class=k>)</span>/:/data2/,/data/RBL_NCI/:/data/RBL_NCI/<span class=w> </span>leafcutter.sif
</code></pre></div></p> <p>Cluster juncfiles <div class=highlight><pre><span></span><code>python<span class=w> </span>leafcutter/clustering/leafcutter_cluster_regtools.py<span class=w> </span>-j<span class=w> </span>juncfiles.txt<span class=w> </span>-m<span class=w> </span><span class=m>50</span><span class=w> </span>-o<span class=w> </span>Experiment<span class=w> </span>-l<span class=w> </span><span class=m>500000</span>
</code></pre></div></p> <p>Prep annotation for leafcutter <div class=highlight><pre><span></span><code>leafcutter/leafviz/gtf2leafcutter.pl<span class=w> </span>-o<span class=w> </span>gencode.v44<span class=w> </span>gencode.v44.annotation.gtf
</code></pre></div></p> <p>Create groups_file.txt. Example: <div class=highlight><pre><span></span><code>Sample1_Aligned.sorted.bam WT
Sample2_Aligned.sorted.bam WT
Sample3_Aligned.sorted.bam WT
Sample4_Aligned.sorted.bam WT
Sample5_Aligned.sorted.bam KO
Sample6_Aligned.sorted.bam KO
Sample7_Aligned.sorted.bam KO
Sample8_Aligned.sorted.bam KO
</code></pre></div></p> <p>Run leafcutter: <div class=highlight><pre><span></span><code>Rscript<span class=w> </span>leafcutter/scripts/leafcutter_ds.R<span class=w> </span>--num_threads<span class=w> </span><span class=m>64</span><span class=w> </span>Experiment_perind_numers.counts.gz<span class=w> </span>groups_file.txt<span class=w> </span>--min_samples_per_intron<span class=w> </span><span class=m>4</span><span class=w> </span>--min_samples_per_group<span class=w> </span><span class=m>4</span><span class=w> </span>-o<span class=w> </span>Experiment
</code></pre></div></p> <p>Create RData file for leafviz: <div class=highlight><pre><span></span><code>Rscript<span class=w> </span>../../leafcutter/leafviz/prepare_results.R<span class=w> </span>--meta_data_file<span class=w> </span>groups_file.txt<span class=w> </span><span class=se>\</span>
<span class=w>  </span>--code<span class=w> </span>leafcutter<span class=w> </span>Experiment_perind_numers.counts.gz<span class=w> </span><span class=se>\</span>
<span class=w>  </span>Experiment_cluster_significance.txt<span class=w> </span><span class=se>\</span>
<span class=w>  </span>Experiment_effect_sizes.txt<span class=w> </span><span class=se>\</span>
<span class=w>  </span>annotation_codes/gencode_v44/gencode.v44<span class=w> </span><span class=se>\</span>
<span class=w>  </span>-o<span class=w> </span>Experiment.RData
</code></pre></div></p> <p>Exit container. Run leafviz locally (for instance with RStudio) <div class=highlight><pre><span></span><code><span class=nf>options</span><span class=p>(</span><span class=n>shiny.host</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=s>&quot;0.0.0.0&quot;</span><span class=p>)</span>
<span class=nf>options</span><span class=p>(</span><span class=n>shiny.port</span><span class=w> </span><span class=o>=</span><span class=w> </span><span class=m>3838</span><span class=p>)</span>
<span class=nf>library</span><span class=p>(</span><span class=n>leafviz</span><span class=p>)</span>
<span class=nf>leafviz</span><span class=p>(</span><span class=s>&quot;~/Lab_Work/CCRRBL-5/leafviz/Experiment.RData&quot;</span><span class=p>)</span>
</code></pre></div></p> </div> </div><script async src="https://www.googletagmanager.com/gtag/js?id=G-D3SL9V30KL"></script> <script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-D3SL9V30KL'); // Replace with your tracking ID
</script> </div> </div> </section> </div> <div class=rst-versions role=note aria-label=Versions> <span class=rst-current-version data-toggle=rst-current-version> <span> <a href=https://github.com/NCI-RBL/Dockers class="fa fa-code-fork" style="color: #fcfcfc"> NCI-RBL/Dockers</a> </span> <span><a href=../init/ style="color: #fcfcfc">&laquo; Previous</a></span> </span> </div> <script src=../../js/jquery-3.6.0.min.js></script> <script>var base_url = "../..";</script> <script src=../../js/theme_extra.js></script> <script src=../../js/theme.js></script> <script src=../../search/main.js></script> <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script> </body> </html> 