---
title: "Peak Annotation Report"
output: html_document

params:
  samplename: ""
  peak_in: ""
  output_table: ""
  readdepth: 3
  PeakIdnt: "ALL" #MultiMap, Unique, all
---


 <!-- # Load libraries, set theme -->
```{r SetEnvironment, include=F ,echo=F,warning=F,message=FALSE}
rm(list=setdiff(ls(), "params"))
library(rmarkdown)
library(stringr)
library(rtracklayer)
library(VariantAnnotation,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(rmarkdown)
library(ggplot2,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library("viridis",quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(edgeR,quietly = T,verbose = F)
library('GenomicFeatures',quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library('rtracklayer',quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(matrixStats,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(plyr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(tidyr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(fitdistrplus,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(stringr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F)
library(data.table)
library(reshape)
library(knitr)
library(stringi)
library(BSgenome)
library(biomaRt)
library(plotly)
library(tidyr)
library(GenomicRanges)
library(RColorBrewer)
library('gplots')
library(ggpubr)
library(circlize)
library('regioneR')
library(scales)
blank_theme <- theme_minimal()+
  theme(
  axis.title.x = element_blank(),
  axis.title.y = element_blank(),
  panel.border = element_blank(),
  panel.grid=element_blank(),
  axis.ticks = element_blank(),
  plot.title=element_text(size=14, face="bold")
  )
```



<!-- # Set params, output   -->

```{r SetParams, echo=F, message=FALSE, warning=FALSE, eval=T, include=F,results=F,}
peak_in=read.table(params$peak_in, sep = '\t', header=TRUE, fill=TRUE)
samplename=params$samplename
readdepth=as.numeric(params$readdepth)
output_table=params$output_table
PeakIdnt=params$PeakIdnt
incd_rRNA=T # What it this?
```

<!-- # Set color schemes  -->

```{r SetColorSchemes, echo=F, message=FALSE, warning=FALSE, eval=T, include=F,results=F,}
nc_color= c(
  'yRNA'= "#A6CEE3",
  'snRNA'= "#1F78B4",
  'snoRNA'="#B2DF8A" ,
  'srpRNA'= "#33A02C",
  'tRNA'=   "#FB9A99",
  '7SKRNA'= "#E31A1C",
  'scRNA'=  "#FDBF6F",
  'sncRNA'= "#FF7F00",
  'scaRNA'= "#CAB2D6",
  'miRNA'= "#6A3D9A" ,
  'vaultRNA'= "#FFFF99",
  'sRNA'= "#B15928"   ,
  'rRNA'=  "#1fb487"  ,
  'miscRNA'= "grey",
  'misc_RNA'= "grey"
)
class_color=c(
  "ncRNA"="#1B9E77",
  "rRNA" ="#1fb487",
  "yRNA"= "#2bdba6" ,
  "protein_coding: Exonic"="#D95F02",
  "Repeat Element"="#7570B3" ,
  "pseudogene"="#E7298A", 
  "Antisense Feature"="#66A61E", 
  "protein_coding: Intronic"="#E6AB02", 
  "lncRNA: Exonic"="#A6761D",
  "lncRNA: Intronic"="#A6761D",
  "no Feature"="#666666")
```

<!-- #   -->



```{r ReadClasses, echo=F, message=FALSE, warning=FALSE, eval=T, include=F,results=F,}
# Identification of CLIP peaks: `r toString(params$samplename)`    

Peaks_Anno = peak_in

Peaks_Anno$Same_Comb_type = 
    factor(peak_in$Same_Comb_type, 
           levels = c("ncRNA", "protein_coding: Exonic", "Repeat Element","pseudogene",
                      "Antisense Feature","protein_coding: Intronic","lncRNA: Intronic", "lncRNA: Exonic","no Feature"))  

Peaks_Anno$Oppo_Comb_type = 
    factor(peak_in$Oppo_Comb_type, 
           levels = c("ncRNA", "protein_coding: Exonic", "Repeat Element","pseudogene",
                      "Antisense Feature","protein_coding: Intronic","lncRNA: Intronic", "lncRNA: Exonic","no Feature"))  


###### remove chrM clip peaks
Peaks_Anno=Peaks_Anno[Peaks_Anno$chrom%in%'chrM'==F,] 
#Peaks_Anno=Peaks_Anno[1:10,]
```



# CLIP peak annotations: `r toString(samplename)`  
  
## Processing data
**1. STAR: align reads**

**2. UMI-tools: remove duplicates**

**3. CTK: call CLIP peaks**

<!-- # Histogram of Reads per Peak  -->


  Identify peaks with All non-spliced reads (unique and Multimapping) and count number of reads that overlap with the identified peaks.
  
  Peaks with > `r toString(readdepth)` (unique + Multi-Mapped:fraction) reads were assessed.  


```{r PeakDepthHist,fig.align='center',echo=F,warning=F,eval=T,include=T}
#print(peak_in)
##todo - make function with replacement title
if (params$PeakIdnt=='UNIQUE'){
  x = ggplot(peak_in, aes(x=Counts_Unique)) +
    geom_histogram(bins = 30) + 
    scale_x_continuous(trans = "log2") + 
    theme_classic() +
    ylab("# of peaks") + 
    xlab("Peak Counts (unique)") + 
    ggtitle("Distribution of peak counts")
} else if (params$PeakIdnt=='ALL'){
  x = ggplot(peak_in, aes(x=Counts_fracMM)) +
    geom_histogram(bins = 30) +
    scale_x_continuous(trans = "log2") + 
    theme_classic() +
    ylab("# of peaks") + 
    xlab("Peak Counts (Unique + FracMM)") + 
    ggtitle("Distribution of peak counts")
}
ggplotly(x,tooltip = c("x", "y"))
```


## Identify CLIP peak Gene Location  

### Identify if CLIP peak overlaps with Intron or Exonic region   

<br>  

<!-- Intron coordinates were calculated from GTF file. -->

<!-- CHECK THIS!!!! A second column was added to identify if the peak also overlapped with the 5'UTR 3'UTR or CDS (Column: Featrue 2) --> 

<br> 

Peaks were annotated with overlapping gene/transcript  

Peaks were annotated by whether they overlap with Host gene intron/exon region  

<br>   

### Identify CLIP peak in ncRNA   


```{r NCRNA? ,echo=F,results='asis',eval=F}
print('SKIP')
# NO IDEA WHAT THIS DOES
#Classification_ncRNA=read.delim(paste0(OUT,"ncRNA_Annotations.txt"), header=T, sep="\t",stringsAsFactors = F)
#kable(Classification_ncRNA[,1:3])
```

<br>  


### Identify peaks in repeat regions   

   Annotate all repeat regions/Classes identified in Repeatmasker Annotation file (UCSC Table browser)  
   Data was not filtered based on any of the identified Repeats.  
    1) LINE/SINE   
    2) LTR   
    3) DNA   
    4) Satellites   
    5) Simple Repeats   
    6) Low Complexity   
    7) Other   
    8) Unknown   
    <br>   


```{r PeaksByRegion, echo=F,warning=F, include=T,eval=T,fig.align='center'}

Repeats_df <- data.frame(matrix(ncol = 3, nrow = 0))
colnames(Repeats_df) <- c('SameFrequency', 'OppoFrequency', 'Repeat_Type')

for (type in c('LINE', 'SINE', 'DNA', 'Low_Complexity', 'LTR', 'Satellites', 'Simple_Repeats', 'Other', 'Unknown')) {
  Repeats_df[nrow(Repeats_df) + 1,] <- c(
  sum(grepl(type, peak_in$Same_Repeat, ignore.case = TRUE)) / nrow(peak_in) *100,
  sum(grepl(type, peak_in$Oppo_Repeat, ignore.case = TRUE)) / nrow(peak_in) *100,
  type)
}

Repeats_df$SameFrequency <- round(as.numeric(Repeats_df$SameFrequency), digits=2)
Repeats_df$OppoFrequency <- round(as.numeric(Repeats_df$OppoFrequency), digits=2)

pSame <- ggplot(Repeats_df) +
  geom_bar(aes(x=Repeat_Type,y=SameFrequency), stat="identity") +
  theme_classic() +
  xlab("") +
  ylab("Frequency of peaks in \nrepeat regions (%)") +
  ggtitle("Same Strand") +
  theme(
    axis.title=element_text(size=15),
    axis.text=element_text(size=10),
    legend.text = element_text( size=10)) +
  theme(axis.text.x = element_text(angle = -45))
ggplotly(pSame,tooltip = c("x", "y"))

pOppo <- ggplot(Repeats_df) +
  geom_bar(aes(x=Repeat_Type,y=OppoFrequency), stat="identity") +
  theme_classic() +
  xlab("") +
  ylab("Frequency of peaks in \nrepeat regions (%)") +
  ggtitle("Opposite Strand") +
  theme(
    axis.title=element_text(size=15),
    axis.text=element_text(size=10),
    legend.text = element_text( size=10)) +
  theme(axis.text.x = element_text(angle = -45))
ggplotly(pOppo,tooltip = c("x", "y"))

```



<br>   
<br>   

### Asigning Clip peak attributes   

Not all Peaks overlap with a single feature so peak assignments were assigned by priority:  

<!-- **ncRNA > Protein coding : Exonic > repeats > Pseudogene > Antisense Feature > Protein Coding : Intronic > lncRNA > no Feature**    -->

**ncRNA > protein_coding: Exonic > Repeat Element > pseudogene > lncRNA: Exonic > Antisense Feature > protein_coding: Intronic > lncRNA: Intronic > no Feature**

All annotations from RNA type, Repeat regions, and Intronic/exonic regions are annotated in the Table.   

<br>  
<br>  


## Overview of targets by CLIP peak category   


```{r PeakCategories, echo=F, include=T, eval=T, fig.align='default', fig.ncol=2, fig.height=3.5, fig.width=7, out.height=200, out.width="50%", fig.show="hold", layout='l-screen', results = FALSE, message=FALSE , warning=F}


#Plot 1
if (params$PeakIdnt=='UNIQUE'){
  x1= ggplot(Peaks_Anno,
            aes(x=Counts_Unique, 
                y=Length,
                color=Same_Comb_type,
                text=Same_external_gene_name)) +
      geom_point(size=1) +
      # scale_x_continuous(trans = "log2",limits = c(.5,NA))+
      scale_y_continuous(trans = "log2",
                          limits = c(NA,NA),
                          oob = squish_infinite) +
      theme_classic() +
      ylab("Peak Width") +
      xlab("Peak Counts (unique)") +
      ggtitle("Comparison of peak counts and Peak Width") +
      scale_color_manual(values=class_color) 

} else if (params$PeakIdnt=='ALL'){
  x1= ggplot(Peaks_Anno,
             aes(x=Counts_fracMM,
                 y=Length,
                 color=Oppo_Comb_type,
                 text=Same_external_gene_name)) +
      geom_point(size=1)+
      # scale_x_continuous(trans = "log2",limits = c(4,NA))+
      scale_y_continuous(trans = "log2",
                          limits = c(NA,NA),
                          oob = squish_infinite) + 
      theme_classic() +
      ylab("Peak Width") +
      xlab("Peak Counts (Unique+ FracMM)") +
      ggtitle("Comparison of peak counts and Peak Width") +
      labs(color='RNA Type') +
      scale_color_manual(values=class_color) 
}

#Plot 2
x2= ggplot(Peaks_Anno,
           aes(x=Length)) +
    geom_histogram(bins = 100) +
    # scale_x_continuous(trans = "log2",limits = c(20,NA)) + 
    theme_classic() +
    ylab("# of peaks") +
    xlab("Peak Width") +
    ggtitle("Distribution of peak Width")

#Plot 3
x3= ggplot(Peaks_Anno,
           aes(x=Length,
               line=Oppo_Comb_type,
               color=Oppo_Comb_type)) +
    geom_density(size=.5) +
    # scale_x_continuous(trans = "log2",limits = c(8,NA)) + 
    theme_classic() +
    ylab("Density") +
    xlab("Peak Width") +
    ggtitle("Distribution of peak Width") +
    labs(color='RNA Type') +
    scale_color_manual(values=class_color) 

#plot 4
x4= ggplot(Peaks_Anno,
           aes(x=Length,
               line=Oppo_Comb_type,
               color=Oppo_Comb_type)) +
    geom_freqpoly(bins = 100)+
    # scale_x_continuous(trans = "log2",limits = c(8,NA)) + 
    theme_classic() +
    ylab("# of peaks") +
    xlab("Peak Width") +
    ggtitle("Distribution of peak Width") +
    labs(color='RNA Type') +
    scale_color_manual(values=class_color) 

#plot 5
if (params$PeakIdnt=='UNIQUE'){
  x5= ggplot(Peaks_Anno,
            aes(x=Counts_Unique,
                line=Oppo_Comb_type,
                color=Oppo_Comb_type)) +
      geom_density() +
      # scale_x_continuous(trans = "log2",limits = c(.25,NA)) + 
      theme_classic() +
      ylab("Density") +
      xlab("Peak Counts (unique)") +
      ggtitle("Distribution of peak counts") + 
      labs(color='RNA Type') +
      scale_color_manual(values=class_color)

} else if (params$PeakIdnt=='ALL'){
  x5= ggplot(Peaks_Anno,
             aes(x=Counts_fracMM,
                 line=Oppo_Comb_type,
                 color=Oppo_Comb_type)) +
      geom_density() +
      # scale_x_continuous(trans = "log2",limits = c(.25,NA)) + 
      theme_classic() +
      ylab("Density") +
      xlab("Peak Counts (Unique+ FracMM)") +
      ggtitle("Distribution of peak counts") +
      labs(color='RNA Type') +
      scale_color_manual(values=class_color) 
}

#Plot all graphs
if (nrow(Peaks_Anno)>1) {
  x1=x1+scale_x_continuous(trans = "log2",limits = c(4,NA));  
  x2=x2+scale_x_continuous(trans = "log2",limits = c(20,NA));  
  x3=x3+scale_x_continuous(trans = "log2",limits = c(8,NA));
  x4=x4+scale_x_continuous(trans = "log2",limits = c(8,NA));  
  x5=x5+scale_x_continuous(trans = "log2",limits = c(.25,NA))
}
  x1;x2;x3;x4;x5
```

<!-- Does this do anything? -->

```{r Identify_length_Issues, echo=F,warning=F,eval=F,include=T}

#############################
### what is the most frequent peak width
#############################
  cnt=count(Peaks_Anno$Length)
  nrow(cnt[cnt$x<20,])
  cnt[order(cnt$freq,decreasing = T),]

    View(Peaks_Anno[Peaks_Anno$Length%in%87,c('ID','IDmerge','Length')])

  Peaks_Anno[Peaks_Anno$Length==2,c('ID','IDmerge')]
  Peaks_Anno[Peaks_Anno$Length==2,c('ID','IDmerge')]
  
count(Peaks_Anno$chr)

Peaks_Anno[grep("exon",Peaks_Anno$Comb_type_exon_Oppo),]
#############################
### Length of ID vs Length reported
#############################
## Lenth reported is sum of splice peak widths
  Peaks_AnnoSPLICE=
  Peaks_Anno[!is.na(Peaks_Anno$IDmerge),c('ID','IDmerge','start','end','Length')]%>%
    separate(col=ID,into=c('ID2','STRAND'),sep="_",remove = F)%>%
    separate(col=ID2,into=c('CHR','START','END'),sep=":|-",remove = F)
  
  Peaks_AnnoSPLICE$START=as.numeric(Peaks_AnnoSPLICE$START)
  Peaks_AnnoSPLICE$END=as.numeric(Peaks_AnnoSPLICE$END)
  

  Peaks_AnnoSPLICE$length2=Peaks_AnnoSPLICE$END-Peaks_AnnoSPLICE$START
```


<br>  
<br>   

##  Number of peaks assigned to each category  


**Peak counts do include rRNA peaks**   

```{r fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}
p=Peaks_Anno
p$PlotColumn=p$Oppo_Comb_type
if (incd_rRNA==FALSE) {
  p=p[!p$Same_Comb_type_ncRNA%in%'rRNA',]
}
if (nrow(p)>0) {
#change types
p[is.na(p$PlotColumn),'PlotColumn']='no Feature'
colnames(p)[colnames(p)%in%'PlotColumn']='RNA_Type'
p$RNA=''
#plot
gg=ggplot(p,aes(fill=RNA_Type,x=RNA) ) +
  geom_bar(position="stack",width=.5) +
  theme_classic()+ggtitle("BioType: # of peaks per category") +
  scale_fill_manual(values=class_color) 
ggplotly(gg)
}else{print("No ncRNA")}
```

