---
title: "MANORM Annotation Report"
output: html_document

params:
  peak_in: ""
  PeakIdnt: ""
  samplename: ""
  background: ""
  pval: .005
  FC: 1
  incd_rRNA: T
---
<!-- # Load libraries, set theme   -->
```{r environment, include=F ,echo=F,warning=F,message=FALSE}
# rm(list=setdiff(ls(), "params"))
suppressMessages(library(rmarkdown))
suppressMessages(library(stringr))
suppressMessages(library(VariantAnnotation,quietly = T,verbose = F,warn.conflicts = F,logical.return = F))
suppressMessages(library(rmarkdown))
suppressMessages(library(ggplot2,quietly = T,verbose = F,warn.conflicts = F,logical.return = F))
suppressMessages(library("viridis",quietly = T,verbose = F,warn.conflicts = F,logical.return = F))
suppressMessages(library(edgeR,quietly = T,verbose = F))
suppressMessages(library('GenomicFeatures',quietly = T,verbose = F,warn.conflicts = F,logical.return = F))
suppressMessages(library(matrixStats,quietly = T,verbose = F,warn.conflicts = F,logical.return = F))
suppressMessages(library(plyr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F))
suppressMessages(library(tidyr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F))
suppressMessages(library(fitdistrplus,quietly = T,verbose = F,warn.conflicts = F,logical.return = F))
suppressMessages(library(stringr,quietly = T,verbose = F,warn.conflicts = F,logical.return = F))
suppressMessages(library(data.table))
suppressMessages(library(reshape))
suppressMessages(library(knitr))
suppressMessages(library(stringi))
suppressMessages(library(BSgenome))
suppressMessages(library(biomaRt))
suppressMessages(library(plotly))
suppressMessages(library(tidyr))
suppressMessages(library(GenomicRanges))
suppressMessages(library(RColorBrewer))
suppressMessages(library('gplots'))
suppressMessages(library(ggpubr))
suppressMessages(library(circlize))
suppressMessages(library('regioneR'))
suppressMessages(library(scales))
suppressMessages(library(dplyr))

blank_theme <- theme_minimal()+
  theme(
  axis.title.x = element_blank(),
  axis.title.y = element_blank(),
  panel.border = element_blank(),
  panel.grid=element_blank(),
  axis.ticks = element_blank(),
  plot.title=element_text(size=14, face="bold")
  )
```

<!-- # Set params, output   -->

```{r echo=F, message=FALSE, warning=FALSE, eval=T, include=F,results=F,}
peak_in=read.table(params$peak_in, sep = '\t', header=TRUE, fill=TRUE)
samplename=params$samplename
background=params$background
pval_Cut=as.numeric(params$pval)
FC_cut=as.numeric(params$FC)
incd_rRNA=params$incd_rRNA
```

```{r echo=F, message=FALSE, warning=FALSE, eval=T, include=F,results=F,}
nc_color= c('yRNA'= "#A6CEE3",
  'snRNA'= "#1F78B4",
  'snoRNA'="#B2DF8A" ,
  'srpRNA'= "#33A02C",
  'tRNA'=   "#FB9A99",
  '7SKRNA'= "#E31A1C",
  'scRNA'=  "#FDBF6F",
  'other_scRNA'=  "#FDBF6F",
  'sncRNA'= "#FF7F00",
  'scaRNA'= "#CAB2D6",
  'miRNA'= "#6A3D9A" ,
  'vaultRNA'= "#FFFF99",
  'sRNA'= "#B15928"   ,
  'rRNA'=  "#1fb487"  ,
  'miscRNA'= "grey",
  'misc_RNA'= "grey")

class_color=c(
  "ncRNA"="#1B9E77",
   "rRNA" ="#1fb487",
   "yRNA"= "#2bdba6" ,
   "protein_coding: exon"="#D95F02",
   "Repeat Element"="#7570B3" ,
   "pseudogene"="#E7298A", 
   "Antisense Feature"="#66A61E", 
   "protein_coding: Intron"="#E6AB02", 
   "lncRNA"="#A6761D", 
   "no Feature"="#666666")
```

```{r echo=F, message=FALSE, warning=FALSE, eval=T, include=T, results=F,}
# Identification of CLIP peaks: `r toString(params$samplename)`    

Peaks_Anno = peak_in

Peaks_Anno$Same_Comb_type = 
    factor(peak_in$Same_Comb_type, 
           levels = c("ncRNA", "protein_coding: Exonic", "Repeat Element","pseudogene",
                      "Antisense Feature", "protein_coding: Intronic", "lncRNA: Intronic", "lncRNA: Exonic","no Feature"))  

Peaks_Anno$Oppo_Comb_type = 
    factor(peak_in$Oppo_Comb_type, 
           levels = c("ncRNA", "protein_coding: Exonic", "Repeat Element","pseudogene",
                      "Antisense Feature", "protein_coding: Intronic", "lncRNA: Intronic", "lncRNA: Exonic","no Feature"))  

###### remove chrM clip peaks
Peaks_Anno=Peaks_Anno[Peaks_Anno$chrom%in%'chrM'==F,] 


#### Update ncRNA categories

Peaks_Anno$Same_ncRNA[grep("HY", Peaks_Anno$Same_ncRNA)] <- 'yRNA'
Peaks_Anno$Same_ncRNA[grep("tRNA", Peaks_Anno$Same_ncRNA)] <- 'tRNA'
Peaks_Anno$Same_ncRNA[grep("7SK", Peaks_Anno$Same_ncRNA)] <- '7SKRNA'
Peaks_Anno$Same_ncRNA <-  gsub('miRNA,','',Peaks_Anno$Same_ncRNA)
Peaks_Anno$Same_ncRNA <-  gsub(' miRNA','',Peaks_Anno$Same_ncRNA)
Peaks_Anno$Same_ncRNA <-  gsub('misc_RNA,','',Peaks_Anno$Same_ncRNA)
Peaks_Anno$Same_ncRNA <-  gsub(' misc_RNA','',Peaks_Anno$Same_ncRNA)
Peaks_Anno$Same_ncRNA <-  gsub(' ','',Peaks_Anno$Same_ncRNA)
Peaks_Anno$Same_ncRNA <-  gsub(',','',Peaks_Anno$Same_ncRNA)

Peaks_Anno$Same_ncRNA[grep("BK000964", Peaks_Anno$chrom)] <- 'rRNA'
Peaks_Anno=Peaks_Anno[!Peaks_Anno$chrom%in%'GL',]
Peaks_Anno=Peaks_Anno[!Peaks_Anno$chrom%in%'JH',]
Peaks_Anno=Peaks_Anno[!Peaks_Anno$chrom%in%'K',]
```


<!-- ## Processing data    -->
# Differential CLIP peaks identified by MAnorm (`r samplename`-`r background`)  

MAnorm used to identify Peaks preferentially expressed in the `r (samplename)` samples compared to `r (background)`.       

## I. Identify preferentialy expressed CLIP peaks  
```{r ,fig.asp=.5, fig.align='center',fig.height=3.5, fig.width=7,echo=F,warning=F,eval=T,include=T}
p=Peaks_Anno

#p$Same_ncRNA[grep("rRNA", p$Same_ncRNA)] <- 'rRNA'

p$Same_Comb_type=factor(p$Same_Comb_type, 
                                        levels = c("ncRNA","rRNA", "protein_coding: exon", 
                                                    "Repeat Element","pseudogene",
                                                    "protein_coding: Intron","lncRNA","no Feature"))

#print(head(p))

p[(p$Same_Comb_type=='ncRNA'&p$Same_ncRNA=='rRNA'),
  'Same_Comb_type']='rRNA'

g1=ggplot(p,aes(x=m_value,y=-log10(p_value),color=`Same_Comb_type`)) +
  geom_point(size=0.5)+
  theme_classic()+
  xlab(paste0("log2 M-value \n (",samplename,"-",background,")"))+
  ylab(paste0("-log10 P-Value"))+
  labs(color='RNA Type')+
  geom_hline(yintercept=-log10(pval_Cut),size=.5,col='black',linetype="dashed")+
  geom_vline(xintercept=(FC_cut),size=.5,col='black',linetype="dashed")+
  geom_vline(xintercept=(-FC_cut),size=.5,col='black',linetype="dashed")+
  scale_color_manual(values=class_color)

g2=ggplot(p,aes(x=log2(p[,'normalized_read_density_sample']),y=m_value,color=-log10(p_value)))+
  geom_point(size=0.5)+
  theme_classic()+
  xlab(paste0("log2 normalized Counts \n (",samplename,")"))+
  ylab(paste0("log2 M-value \n(",samplename,"-",background,")"))+
  scale_colour_gradient(  low = ("blue"),  high = ("red"),
                           na.value = "grey50",guide = "colourbar")+
  geom_hline(yintercept=(FC_cut),size=.5,col='black',linetype="dashed")+
  geom_hline(yintercept=(-FC_cut),size=.5,col='black',linetype="dashed")
    
g3=ggplot(p,aes(x=log2(p[,'normalized_read_density_sample']),y=m_value,color=`Same_Comb_type`))+
  geom_point(size=0.5)+
  theme_classic()+
  xlab(paste0("log2 normalized Counts \n (",samplename,")"))+
  ylab(paste0("log2 M-value \n(",samplename,"-",background,")"))+
  labs(color='RNA Type')+
  geom_hline(yintercept=(FC_cut),size=.5,col='black',linetype="dashed")+
  geom_hline(yintercept=(-FC_cut),size=.5,col='black',linetype="dashed")+
  scale_color_manual(values=class_color) 
if (nrow(p)<5000) {
  g1=ggplotly(g1,tooltip = c("text"))
  g2=ggplotly(g2,tooltip = c("text"))
  g3=ggplotly(g3,tooltip = c("text"))
  }
g1;g2;g3
```

```{r ,fig.asp=.5, fig.align='center',echo=F,warning=F,eval=T,include=T}
MPeak=median(Peaks_Anno$end - Peaks_Anno$start)
```

```{r fig.asp=.1, fig.align='center',fig.height=.1,echo=F,warning=F,eval=T,include=T}

MPeakPrint=paste0("Median Peak width: ",MPeak,"nt\n")

if (MPeak>=50&MPeak<=100) {
matrix(1:9, nrow = 3, ncol = 3)%>%as.data.frame()%>%ggplot(.,aes(x=V1,y=V2))+geom_text(x=.4, y=.1, label=MPeakPrint,size=4.5)+theme_void()
}
```

```{r fig.asp=.5, fig.align='center', fig.width=5,echo=F,warning=F,eval=T,include=T}
if (MPeak<50|MPeak>100) {
  
txt1=paste0( "Consider modifying MAnorm values\n",
                             "-w ",MPeak," and -d ",MPeak/2
                                                          )
txt2=paste0(
                             "Default Values:\n",
                             "-w 50 and -d 25"
                             )
  
matrix(1:3, nrow = 3, ncol = 3)%>%as.data.frame()%>%ggplot(.,aes(x=V1,y=V2))+
  geom_text(x=.4, y=.8, label=MPeakPrint,size=6,color='red')+
  geom_text(x=.4, y=.55, label=txt1,size=6,color='red')+
  geom_text(x=.4, y=.2, label=txt2,size=5,color='black')+
  theme_void()

}

```

<br>  

```{r ,fig.asp=.5, fig.align='center',echo=F,warning=F,eval=T,include=T}

Peaks_Anno=Peaks_Anno[Peaks_Anno$p_value<pval_Cut&abs(Peaks_Anno$m_value)>FC_cut,]
```

## Peaks with a P-value < `r pval_Cut` and |log~2~(fold change)| > `r FC_cut` were selected, resulting in `r toString(nrow(Peaks_Anno))` peaks.   

<br>



### II.  Number of peaks assigned to each category  


**II.1 Peak counts do include rRNA peaks**   

```{r PeaksPerCategory, fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}
p=Peaks_Anno
p$PlotColumn=p$Same_Comb_type
if (incd_rRNA==FALSE) {
  p=p[!p$Same_Repeat%in%'rRNA',]
}
if (nrow(p)>0) {
#change types
p[is.na(p$PlotColumn),'PlotColumn']='no Feature'
colnames(p)[colnames(p)%in%'PlotColumn']='RNA_Type'
p$RNA=''
#plot
gg=ggplot(p,aes(fill=RNA_Type,x=RNA) ) +
  geom_bar(position="stack",width=.5) +
  theme_classic()+ggtitle("BioType: # of peaks per category") +
  scale_fill_manual(values=class_color) 
ggplotly(gg)
}else{print("No ncRNA")}
```


```{r PeaksPer_NCRNA_Category, fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}
p=Peaks_Anno
p$PlotColumn=p$Same_Comb_type

###phil need else statement?###
#check if this makes sense?
if (incd_rRNA==FALSE) {
  p=p[!p$Same_ncRNA%in%'rRNA',]
} else{
  p=p
}
###############################

#filter
p=p[p$PlotColumn%in%'ncRNA','Same_ncRNA',drop=F]

#if there are remaining types
if (nrow(p)>0) {
  if(length(unique(p$Same_ncRNA)[unique(p$Same_ncRNA)%in%names(nc_color)==F])>0){
    print(paste0('unexpected ncRNA category: ',unique(p$Same_ncRNA)[unique(p$Same_ncRNA)%in%names(nc_color)==F]))
  }else{
    nc_color_sel=nc_color[unique(p$Same_ncRNA)]
    nc_color_sel=nc_color_sel[is.na(nc_color_sel)==F]  
    p$RNA=""
    gg=ggplot(p,aes(fill=Same_ncRNA,x=RNA) ) +
        geom_bar(width=.5) +
        theme_classic() +
        ggtitle("ncRNA only: # of peaks per category") +
        labs(fill='RNA Type') +
        scale_color_manual(values=nc_color_sel) +
        scale_fill_manual(values=nc_color_sel) + 
        blank_theme + 
        coord_polar("y",start=0)
    (gg) ###updating print to be consistent
    #ggplotly(gg)
  }  
}else{print("No ncRNA")}
```

<br>  
<br> 

**II.2 Peak counts do not include rRNA peaks**   


```{r PeaksPerCategory_NO_rRNA, fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}
p=Peaks_Anno
p$PlotColumn=p$Same_Comb_type
p=p[!p$Same_ncRNA%in%'rRNA',]
if (nrow(p)>0) {
p[is.na(p$Same_Comb_type),'PlotColumn']='no Feature'
colnames(p)[colnames(p)%in%'PlotColumn']='RNA_Type'
p$RNA=''
#plot
gg=ggplot(p,aes(fill=RNA_Type,x=RNA) )+
  geom_bar(position="stack",width=.5) +
  theme_classic() +
  ggtitle("BioType: # of peaks per category") +
  scale_fill_manual(values=class_color) 
ggplotly(gg)
}else{print("No ncRNA")}
```


```{r PeaksPer_NCRNA_Category_NO_rRNA, fig.asp=.5, fig.align='center',fig.height=4, fig.width=8,echo=F,warning=F,eval=T,include=T}
p=Peaks_Anno
p$PlotColumn=p$Same_Comb_type
p=p[!p$Same_ncRNA%in%'rRNA',]
p=p[p$PlotColumn%in%'ncRNA','Same_ncRNA',drop=F]
#if there are remaining types
if (nrow(p)>0) {
  if(length(unique(p$Same_ncRNA)[unique(p$Same_ncRNA)%in%names(nc_color)==F])>0){
    print(paste0('unexpected ncRNA category: ',unique(p$Same_ncRNA)[unique(p$Same_ncRNA)%in%names(nc_color)==F]))
}else{
  nc_color_sel=nc_color[unique(p$Same_ncRNA)]
  nc_color_sel=nc_color_sel[is.na(nc_color_sel)==F]  
    p$RNA=""
    gg=ggplot(p,aes(fill=Same_ncRNA,x=RNA) ) +
      geom_bar(width=.5)+theme_classic() +
      ggtitle("ncRNA only: # of peaks per category") +
      labs(fill='ncRNA Type') +
      scale_color_manual(values=nc_color_sel) + 
      scale_fill_manual(values=nc_color_sel) +
      blank_theme + 
      coord_polar("y",start=0)
    (gg)
    # ggplotly(gg)
}
}else{print("No ncRNA")}
```




### III. Opposite strand CLIP peak attributes   


```{r , fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F}
p=Peaks_Anno
#p=p$Oppo_Comb_type
###phil need else statement
if (incd_rRNA==FALSE) {
  p=p[!p$Oppo_Comb_type%in%'rRNA',]
}
if (nrow(p)>0) {
colnames(p)[colnames(p)%in%'Oppo_Comb_type']='RNA_Type'
p$RNA=''
#plot
gg = ggplot(p,aes(fill=RNA_Type,x=RNA_Type)) +
  geom_bar(stat="count") +
  theme_classic() +
  ggtitle("Opposite Strand Feature - Number of peaks by category") +
  xlab("") +
  theme(plot.title = element_text(hjust = 0.5),
        axis.title = element_text(size=15),
        axis.text = element_text(size=10),
        axis.text.x = element_text(angle = -45,vjust = .5)) +
  scale_fill_manual(values=class_color) 
ggplotly(gg)
}else {print("No Features")}
```


```{r fig.asp=.5, fig.align='center',fig.height=6, fig.width=12,echo=F,warning=F,eval=T,incude=T}
p=Peaks_Anno
##phil need else statement
if (incd_rRNA==FALSE) {
  p=p[!p$Oppo_Comb_type%in%'rRNA',]
}
#p=p[p$Comb_type_exon_Oppo%in%'Antisense Feature',]
if (nrow(p)>0) {
  p$transition=paste0(p$Same_Comb_type,'->\n',p$Oppo_Comb_type)
colnames(p)[colnames(p)%in%'transition']='RNA_Type'
p$Difference=''
#plot
gg = ggplot(p,aes(fill=Same_Comb_type,x=factor(Oppo_Comb_type))) +
      geom_bar(stat="count") +
      theme_classic() +
      ggtitle("Opposite Strand Feature Features: \n Compare Antisense and Sense stand Annotation") +
      xlab("Antisense strand Annotation") +
      ylab("Peak Count") +
      labs(fill = "Sense Stand Annotation") +
      theme(plot.title = element_text(hjust = 0.5,size=20),
            axis.title = element_text(size=15),
            axis.text = element_text(size=10),
            legend.position = "right",
            axis.text.x = element_text(angle = -45,vjust = .5))
ggplotly(gg)
}else {print("No Features")}
```

<br>  